### PR TITLE
Get rid of GoToAdjPos subgoal

### PR DESCRIPTION
**Pull request by saleml**
- Fixes https://github.com/mila-udem/babyai/issues/14
- Code is more factorized (simpler maintenance) : https://github.com/mila-udem/babyai/issues/16
- As this should, in theory, only affects PutNext Level  (and Seq levels, such as BossLevel), I tested it on PutNext and BossLevel:
 - The results of PutNext were similar to the best version (reported in https://github.com/mila-udem/babyai/pull/3).
 - I observed however a slight increase of avg. # steps has increased in BossLevel from 86.66 to 86.92 on 500 levels.

### CODE REVIEW

**Comment by saleml** [comment]
while the sligh decrease of performance on BossLevel doesn't seem to be a real issue, it showed an undesired behavior (e.g. BossLevel seed 462), related to the TODO in https://github.com/mila-udem/babyai/pull/17/files#diff-23560fa0078a677e0990569da3602cd5R552.
@maximecb, is there a quick way to test if two positions are in the same room ?

**Comment by maximecb** [comment]
You can call a method of the room objects to check if some position `(i, j)` is inside the room:

https://github.com/maximecb/gym-minigrid/blob/master/gym_minigrid/roomgrid.py#L47

You'd first have to call `room_from_pos(i, j)` to get the room the agent is in: https://github.com/maximecb/gym-minigrid/blob/master/gym_minigrid/roomgrid.py#L99

File: babyai/bot.py [hunk 1]
```diff
@@ -333,10 +350,20 @@ def get_action(self):
                 # otherwise: forward, left and right are full, we assume that you can go behind
                 return self.actions.left
 
-            # If we are right in front of the object, then the subgoal is completed
-            # go back to the previous subgoal
-            if np.array_equal(obj_pos, self.fwd_pos):
-                return self.subgoal_accomplished()
+            if not self.adjacent:
```
Comment (rizar) [inline babyai/bot.py]: A bit of nit-picking: `if self.adjacent` is better than `if not self.adjacent`:

https://www.python.org/dev/peps/pep-0008/#programming-recommendations

**Comment by saleml** [comment]
Should be good to go now. But @rizar maybe you should regenerate the demos for BossLevel, to  make sure that the first failure (if any) happens AFTER what was the first failure before this PR.

> Looks cool! How come we were able to get rid of the `Update...` subgoal?

UpdateStackSubgoal was used when GoToAdjPos(obj) gave birth to GoNextTo(free cell adj to obj), and while realizing the GoNextTo subgoal, the agent needed to move the obj. Think of the bot having to drop a key (it already holds) next to a blue ball, and the blue ball is right behind a door that just opened.

Now, DropSubgoal automatically updates the stack (whenever you drop something, you loop through each subgoal to see if you were supposed to go to the object you just dropped, in which case, you should update its position). So as soon as unblocking (in the scenario described above) is done, the stack should be fine. Now I basically force unblocking (in that particular scenario) with:

```
# corner case: I just opened a door, and the target is right after the door
                    elif self.fwd_cell.type.endswith('door') and self.fwd_cell.is_open:
                        # add a useless subgoal that will force unblocking
                        return GoNextToSubgoal(self.bot, self.fwd_pos + 2 * self.dir_vec).get_action()
```
It works

**Comment by rizar** [comment]
This is great, I am gonna launch generation of 1M demos right away.

**Comment by maximecb** [comment]
:crossed_fingers:

**Comment by rizar** [comment]
I think we have a regression, `--shift=1093 --seed=1`

**Comment by rizar** [comment]
Turns out, there are cells such that you can go through them but you can't put anything in them. An open door is the only example I have in mind so far. We should therefore looks for paths that don't have a door as the last cell in them. I will try to see how to achieve this.

**Comment by rizar** [comment]
Closing this in favor of #19

**Pull request closed without merging**